### ISSUE TITLE
Relations: If a relation scalar is required, the relation also has to be required

### ISSUE THREAD

**Issue by janpio** [prisma/prisma#20406]
Look at this schema:
```prisma
model User {
  id     String  @id @default(cuid())
  email  String? @unique
  author Author?
}

model Author {
  id    String @id @default(cuid())
  email String @unique
  name  String

  user User? @relation(fields: [email], references: [email])
}
```

`Author.email` is required, but `Author.user` is not. With `email` never being allowed to be null, it is impossible that there is _no_ relation to `User` possible for this model.

We should disallow this in validation so it is clear that an required relations scalar field also requires a required relation.

Note: This might be more complicated for relations with multiple scalar fields.

--- 

- Internal discussion: https://prisma-company.slack.com/archives/C01TM7UM83Y/p1690323875179509
- Triggered by https://github.com/prisma/prisma-engines/pull/4083 + https://github.com/prisma/prisma/issues/18058

**Comment by samualtnorman** [comment]
Is this a duplicate of https://github.com/prisma/prisma/issues/20400?
Disallowing this schema would cause Prisma to lose the ability to model relationships where the optionality is on the relationship rather than the scalar field. This would force some people who are adding Prisma to their existing project to refactor their database (which isn't always possible (especially when existing code relies on the status quoe)).

**Comment by janpio** [comment]
> Is this a duplicate of https://github.com/prisma/prisma/issues/20400?
No, you are reporting a bug that might or might not be valid, and https://github.com/prisma/prisma/issues/20406 is an explicit improvement request. We treat this separately, but as they are not cross linked will probably resolved together. If the bug is handled some way, the improvement request becomes obsolete - or the other way around.

> Disallowing this schema would cause Prisma to lose the ability to model relationships where the optionality is on the relationship rather than the scalar field. This would force some people who are adding Prisma to their existing project to refactor their database

If the scalar field is required, there is no way for the relation to be optional in our opinion. Prisma does not allow invalid-to-the-relation (which is enforced by the required foreign key of a relation) values in the relation scalar field in the relations we currently support.

So in the example above you can not create a `Author` with an `email` that does not already exist in `User`.
If `Author.email` was optional, it would be possible to create an `Author` that does _not_ have an `email` value - but for that both relation scalar and relation field need to be optional.

**Comment by samualtnorman** [comment]
> but have values that do not exist on the other side
what do you mean?

**Comment by janpio** [comment]
> what do you mean?

1) The (required) field `Author.email` can only have a value that also exists in `User.email`, then it is a valid required relation.
2) If it was optional, it could also have the value `null` - and the relation would then be optional.
3) What Prisma (and foreign keys) does not allow, is that `Author.email` has a value that does not exist in `User.email` - the foreign key will throw a constraint error.

I understood you want to be able to use 3) for your legacy database. Was that a misunderstanding?

**Comment by samualtnorman** [comment]
> I understood you want to be able to use 3) for your legacy database.
Uh no, with what I'm talking about, `User.email` and `Author.email` are both required fields.

**Comment by janpio** [comment]
Yes, if they are both required and can only contain values present in the other, you have a required relation and can use `@relation` in Prisma. That is what this issue suggest to make clear. 

If they are both required, but `Author.email` can be a value that does _not_ exist in `User.email`, then you have a different kind of relation (one that can be valie, or be "invalid" - not optional) that Prisma does not support yet.

(Maybe share an example of the data you have, I think that would make it clearer)

**Comment by samualtnorman** [comment]
sure, here's an example
```sql
create table User (
	name varchar(191) not null,
	email varchar(191) not null,
	unique index User_email_key(email)
	-- pretend there are more fields
);

create table Author (
	name varchar(191) not null,
	email varchar(191) not null,
	unique index Author_email_key(email)
	-- pretend there are more fields
);

insert into User values ("Regina Barker", "rbarker@example.com");
insert into User values ("Mark McDaniel", "mmcdaniel@example.com");
insert into Author values ("Mark McDaniel", "mmcdaniel@example.com");
insert into Author values ("Clark Stout", "cstout@example.com");
```
